pkg/scheduler/admission/podnodeconstraints/admission.go

```diff
@@ -0,0 +1,201 @@
+	extensions.Resource("replicasets"):      extensions.Kind("ReplicaSet"),
+	extensions.Resource("jobs"):             extensions.Kind("Job"),
+	deployapi.Resource("deploymentconfigs"): deployapi.Kind("DeploymentConfig"),
+}
+
+func shouldAdmitResource(resource unversioned.GroupResource, kind unversioned.GroupKind) (bool, error) {
+	expectedKind, shouldAdmit := resourcesToAdmit[resource]
+	if !shouldAdmit {
+		return false, nil
+	}
+	if expectedKind != kind {
+		return false, fmt.Errorf("Unexpected resource kind %v for resource %v", &kind, &resource)
+	}
+	return true, nil
+}
+
+// var _ = oadmission.WantsOpenshiftClient(&podNodeConstraints{})
+var _ = oadmission.Validator(&podNodeConstraints{})
+var _ = oadmission.WantsAuthorizer(&podNodeConstraints{})
+
+func readConfig(reader io.Reader) (*api.PodNodeConstraintsConfig, error) {
+	if reader == nil || reflect.ValueOf(reader).IsNil() {
+		return nil, nil
+	}
+
+	obj, err := configlatest.ReadYAML(reader)
+	if err != nil {
+		return nil, err
+	}
+	if obj == nil {
+		return nil, nil
+	}
+	config, ok := obj.(*api.PodNodeConstraintsConfig)
+	if !ok {
+		return nil, fmt.Errorf("unexpected config object: %#v", obj)
+	}
+	// No validation needed since config is just list of strings
+	return config, nil
+}
+
+func (o *podNodeConstraints) Admit(attr admission.Attributes) error {
+	switch {
+	case o.config == nil,
+		attr.GetSubresource() != "":
+		return nil
+	}
+	shouldAdmit, err := shouldAdmitResource(attr.GetResource(), attr.GetKind())
+	if err != nil {
+		return err
+	}
+	if !shouldAdmit {
+		return nil
+	}
+	// Only check Create operation on pods
+	if attr.GetResource() == kapi.Resource("pods") && attr.GetOperation() != admission.Create {
+		return nil
+	}
+	ps, err := o.getPodSpec(attr)
+	if err == nil {
+		return o.admitPodSpec(attr, ps)
+	}
+	return err
+}
+
+// extract the PodSpec from the pod templates for each object we care about
+func (o *podNodeConstraints) getPodSpec(attr admission.Attributes) (kapi.PodSpec, error) {
+	switch r := attr.GetObject().(type) {
+	case *kapi.Pod:
+		return r.Spec, nil
+	case *kapi.ReplicationController:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Deployment:
+		return r.Spec.Template.Spec, nil
+	case *extensions.ReplicaSet:
+		return r.Spec.Template.Spec, nil
+	case *extensions.Job:
+		return r.Spec.Template.Spec, nil
+	case *deployapi.DeploymentConfig:
+		return r.Spec.Template.Spec, nil
+	}
+	return kapi.PodSpec{}, kapierrors.NewInternalError(fmt.Errorf("No PodSpec available for supplied admission attribute"))
+}
+
+// validate PodSpec if NodeName or NodeSelector are specified
+func (o *podNodeConstraints) admitPodSpec(attr admission.Attributes, ps kapi.PodSpec) error {
+	matchingLabels := []string{}
+	// nodeSelector blacklist filter
+	if len(ps.NodeSelector) > 0 {
+		for nodeSelectorLabel := range ps.NodeSelector {
+			for _, blacklistLabel := range o.config.NodeSelectorLabelBlacklist {
```

Comment:
`sets.String.Intersection` is probably easier to read.  Using a `sets.String` for the `NodeSelectorLabelBlacklist` should definitely be done.